ddprof-lib/src/main/cpp/profiler.cpp

```diff
@@ -922,35 +929,44 @@ bool Profiler::crashHandler(int signo, siginfo_t *siginfo, void *ucontext) {
     // Act as if the load returned default_value argument
     frame.pc() += length;
     frame.retval() = frame.arg1();
+    if (thrd != nullptr) {
+      thrd->exitCrashHandler();
+    }
     return true;
   }
 
   if (WX_MEMORY && Trap::isFaultInstruction(pc)) {
+    if (thrd != nullptr) {
+      thrd->exitCrashHandler();
+    }
     return true;
   }
 
   if (VM::isHotspot()) {
     // the following checks require vmstructs and therefore HotSpot
-    StackWalker::checkFault();
+
+    // this check can longjmp to a completely different location - need to call exitCrashHandler() before
+    StackWalker::checkFault(thrd);
 
     // Workaround for JDK-8313796. Setting cstack=dwarf also helps
     if (VMStructs::isInterpretedFrameValidFunc((const void *)pc) &&
         frame.skipFaultInstruction()) {
+      if (thrd != nullptr) {
+        thrd->exitCrashHandler();
+      }
       return true;
     }
   }
 
+  if (thrd != nullptr) {
+    thrd->exitCrashHandler();
+  }
   return false;
 }
 
 void Profiler::setupSignalHandlers() {
-  orig_trapHandler =
-      OS::installSignalHandler(SIGTRAP, Profiler::trapHandlerEntry);
-  if (orig_trapHandler == (void *)SIG_DFL ||
-      orig_trapHandler == (void *)SIG_IGN) {
-    orig_trapHandler = NULL;
-  }
-  if (VM::java_version() > 0) {
+  // do not re-run the signal setup (run only when VM has not been loaded yet)
+  if (VM::java_version() > 0 && !VM::loaded()) {
```

Comment:
I am not sure I understand this.
Would you not expect to chain signal handlers once the VM is loaded (ensuring you do not get overridden following the load of the VM ?).

Reply:
We want to run this on a vm-init callback only. The async-profiler can do 'late initialization' and, theoretically, run this multiple times. While it really does not change anything for us, having the logic in-sync with async-profiler would make it easier to reason about any changes we might cherry-pick